torch/distributed/algorithms/quantization.py

```diff
@@ -0,0 +1,117 @@
+        ):
+            raise RuntimeError(
+                "Invalid function argument. Expected "
+                "to be of type List[torch.Tensor]."
+            )
+        if (qtype == QuantType.FP16):
+            input_tensor = [self.quntize_tensor(t, qtype) for t in tensor_list]
+            return input_tensor
+        else:
+            raise RuntimeError(
+                'Quantization type {} is not supported'.format(qtype)
+            )
+
+    def dequantize_tensor(self, tensor, qtype, quant_loss=None):
+        if not isinstance(tensor, torch.Tensor):
+            raise RuntimeError(
+                "Invalid function argument. Expected "
+                "to be a single tensor list  of type torch.Tensor."
+            )
+        if (qtype == QuantType.FP16):
+            if tensor.dtype != torch.float16:
+                raise RuntimeError(
+                    "tensor dtype is {} while expected to be FP16.".format(
+                        tensor.dtype
+                    )
+                )
+            elif tensor.dtype == torch.float16 and quant_loss is None:
+                return tensor.float()
+            else:
+                return tensor.float() / quant_loss
+        else:
+            raise RuntimeError(
+                'Quantization type {} is not supported'.format(qtype)
+            )
+
+
+    def dequantize_tensor_list(self, tensors, qtype, quant_loss=None):
+        if not isinstance(tensors, list) or not all(
+            isinstance(p, torch.Tensor) for p in tensors
+        ):
+            raise RuntimeError(
+                "Invalid function argument. Expected "
+                "to be of type List[torch.Tensor]."
+            )
+        elif (qtype == QuantType.FP16):
+            tensor = [self.dequantize_tensor(t, qtype) for t in tensors]
+            return tensor
+        else:
+            raise RuntimeError(
+                'Quantization type {} is not supported'.format(qtype)
+            )
+
+
+def auto_quantize(func, qtype, quant_loss=None):
+    def wrapper(tensors, *args, **kwargs):
+        group = kwargs.get('group', None)
+        async_op = kwargs.get('async_op', False)
+        quant = quantization_wrapper()
+        if (func == dist.all_gather):
+            input_tensors = quant.quntize_tensor(args[0], qtype)
+            out_tensors = quant.quntize_tensor_list(tensors, qtype)
+            work = dist.all_gather(out_tensors, input_tensors, group=group, async_op=async_op)
+
+        elif (func == dist.all_to_all):
+            input_tensors = quant.untize_tensor(args[0], qtype)
+            out_tensors = quant.quntize_tensor(tensors, qtype)
+            work = dist.all_to_all(out_tensors, input_tensors, group=group, async_op=async_op)
+
+        else:
+            raise RuntimeError(
+                'The collective ops {} is not supported yet'.format(func)
+            )
+
+        for i, t in enumerate(quant.dequantize_tensor_list(out_tensors, qtype, quant_loss=quant_loss)):
```

Comment:
I think this will only work if it's not an `async_op`, which means `dist.all_to_all` already finished `wait()` and `out_tensors` already filled with real content. 

What if `async_op=True`? In this case `out_tensors` is not ready, and we are dequantizing them preemptively, this is wrong as it makes the dequantization useless. What we need to do in this case is to wait until the `work` called `wait()` and `out_tensors` are ready, then apply dequantization there. 

This change might be more involved (i.e. we probably need to add `callback` functionality to `ProcessGroup::Work` and fire the callback when results is ready. This don't need to happen in this diff, we can address it later. But we need to throw error to the user in this PR to warn that the `async_op=True` mode is not supported yet.